lib/auth/auth_with_roles.go

```diff
@@ -1484,7 +1485,279 @@ func (a *ServerWithRoles) GetNode(ctx context.Context, namespace, name string) (
+}
+
+// ListUnifiedResources returns a paginated list of unified resources filtered by user access.
+func (a *ServerWithRoles) ListUnifiedResources(ctx context.Context, req *proto.ListUnifiedResourcesRequest) (*proto.ListUnifiedResourcesResponse, error) {
+	// Fetch full list of resources in the backend.
+	var elapsedFetch time.Duration
+	var elapsedFilter time.Duration
+	var unifiedResources []types.ResourceWithLabels
+	var filteredResources []types.ResourceWithLabels
+
+	defer func() {
+		log.WithFields(logrus.Fields{
+			"user":           a.context.User.GetName(),
+			"elapsed_fetch":  elapsedFetch,
+			"elapsed_filter": elapsedFilter,
+		}).Debugf(
+			"ListUnifiedResources(%v->%v) in %v.",
+			len(unifiedResources), len(filteredResources), elapsedFetch+elapsedFilter)
+	}()
+
+	startFetch := time.Now()
+	unifiedResources, err := a.authServer.UnifiedResourceCache.GetUnifiedResources(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	elapsedFetch = time.Since(startFetch)
+
+	filteredResources = make([]types.ResourceWithLabels, 0)
+	startFilter := time.Now()
+	for _, resource := range unifiedResources {
+		switch r := resource.(type) {
+		case types.Server:
+			{
+				if err := a.checkAccessToNode(r); err != nil {
+					if trace.IsAccessDenied(err) {
+						continue
+					}
+
+					return nil, trace.Wrap(err)
+				}
+
+				filteredResources = append(filteredResources, resource)
+			}
+		case types.DatabaseServer:
+			{
+				if err := a.checkAccessToDatabase(r.GetDatabase()); err != nil {
+					if trace.IsAccessDenied(err) {
+						continue
+					}
+
+					return nil, trace.Wrap(err)
+				}
+
+				filteredResources = append(filteredResources, resource)
+			}
+
+		case types.AppServer:
+			{
+				if err := a.checkAccessToApp(r.GetApp()); err != nil {
+					if trace.IsAccessDenied(err) {
+						continue
+					}
+
+					return nil, trace.Wrap(err)
+				}
+
+				filteredResources = append(filteredResources, resource)
+			}
+		case types.SAMLIdPServiceProvider:
+			{
+				if err := a.action(apidefaults.Namespace, types.KindSAMLIdPServiceProvider, types.VerbList); err == nil {
+					filteredResources = append(filteredResources, resource)
+				}
+			}
+		case types.KubeServer:
+			kube := r.GetCluster()
+			if err := a.checkAccessToKubeCluster(kube); err != nil {
+				if trace.IsAccessDenied(err) {
+					continue
+				}
+
+				return nil, trace.Wrap(err)
+			}
+
+			filteredResources = append(filteredResources, kube)
+		case types.WindowsDesktop:
+			{
+				if err := a.checkAccessToWindowsDesktop(r); err != nil {
+					if trace.IsAccessDenied(err) {
+						continue
+					}
+
+					return nil, trace.Wrap(err)
+				}
+
+				filteredResources = append(filteredResources, resource)
+			}
+		}
+	}
+	elapsedFilter = time.Since(startFilter)
+
+	if req.SortBy.Field != "" {
+		isDesc := req.SortBy.IsDesc
+		switch req.SortBy.Field {
+		case types.ResourceMetadataName:
+			sort.SliceStable(filteredResources, func(i, j int) bool {
```

Comment:
Do we need a stable sort here? You can probably squeeze out a little more perf without.

Reply:
What if this introduces artifacts on page boundaries? Consider the following scenario:

1. We have resources A and B, both sharing the same name.
2. The client requests a page of results sorted by name. Sorting algorithm puts A before B. A falls outside the first page, but it will be the first one returned next, so we send the response with the "next key" field pointing to the key of resource A.
3. In the meantime, another resource C appears in the system; it doesn't really matter where it's gonna be. All it matters is that it would be returned by our list at some point of the time.
4. The client requests another page, giving A's key as the start key.
5. We perform an unstable sort. Since the data set changed (resource C appeared in the result set), this time our sorting algorithm decides to put B before A.

As a result, our list of resources would never contain B!